test/regress_finalize.c

```diff
@@ -290,6 +292,60 @@ test_fin_within_cb(void *arg)
 	;
 }
 
+static void
+timer_callback_3(evutil_socket_t fd, short what, void *arg)
+{
+	// nop
+}
+
+static void
+event_finalize_callback_3(struct event *ev, void *arg)
+{
+	struct event_and_count *evc = arg;
+
+	if (evc->count == 0) {
+		memset(evc->ev, 0, sizeof(*evc->ev));
+		event_assign(evc->ev, evc->base, -1, EV_TIMEOUT, timer_callback_3, evc);
+		evc->count += 1;
+	} else {
+		free(evc->ev);
+		free(evc);
+	}
+}
+
+static void
+test_fin_debug_use_after_free(void *arg)
+{
+	struct basic_test_data *data = arg;
+	struct event_base *base = data->base;
+	struct timeval tv = {.tv_sec = 4, .tv_usec = 0};
+
+	// Allocate ev outside of libevent
+	struct event_and_count *evc = malloc(sizeof(*evc));
+	memset(evc, 0, sizeof(*evc));
+	evc->ev = malloc(sizeof(*evc->ev));
+	memset(evc->ev, 0, sizeof(*evc->ev));
+	evc->base = base;
+
+	// Assign, add event, and finalize
+	event_assign(evc->ev, base, -1, EV_TIMEOUT, timer_callback_3, evc);
+	event_add(evc->ev, &tv);
+	event_finalize(0, evc->ev, event_finalize_callback_3);
+
+	// Dispatch base to trigger callbacks
+	event_base_dispatch(base);
+	event_base_assert_ok_(base);
+
+	// Now add again, since we did event_assign in event_finalize_callback_3
+	event_add(evc->ev, &tv);
```

Comment:
Actually the unit test will bail on this line without the fix (no below), the reason is that just after `event_assign()` from the callback, `event_debug_note_teardown_()` will be called which will reset the initialized state.

And the documentation for finalizers API says:
> Once the callback has been invoked, you should treat the event structure as
>  containing uninitialized memory.

Plus to overcome this you can use `event_free_finalize()`, which will automatically free event for you.

I'm not against fixing this, I just need more time to think on this unit test.

> there are some caveats that should be addressed before

Also I added `TT_ENABLE_DEBUG_MODE` test flag, and it is definitely required for this test.